### PR TITLE
Add Apa102 direction getter and setter

### PR DESCRIPTION
**Pull request by tomaszrykala**
Addresses https://github.com/androidthings/contrib-drivers/issues/33

### CODE REVIEW

File: rainbowhat/src/main/java/com/google/android/things/contrib/driver/rainbowhat/RainbowHat.java [hunk 1]
```diff
@@ -92,4 +92,8 @@ public static Gpio openLed(@LedPin String pin) throws IOException {
     public static Apa102 openLedStrip() throws IOException {
         return new Apa102(BUS_LEDSTRIP, Apa102.Mode.BGR);
     }
+
+    public static Apa102 openLedStripReversed() throws IOException {
```
Comment (proppy) [inline rainbowhat/src/main/java/com/google/android/things/contrib/driver/rainbowhat/RainbowHat.java]: what about adding an overload of the method with a `Apa102.Direction` parameter?

Comment (tomaszrykala) [inline rainbowhat/src/main/java/com/google/android/things/contrib/driver/rainbowhat/RainbowHat.java]:
```java
public static Apa102 openLedStrip(Apa102.Direction direction) {... }
``` 
?

Given there are only two possible `Direction`s I think an explicit name would be better but can implement your suggestion just the same.

Comment (proppy) [inline rainbowhat/src/main/java/com/google/android/things/contrib/driver/rainbowhat/RainbowHat.java]: yep, I'd personally prefer the explicit direction :) but I'll let other folks more friendly with Android style guide comment.

**Comment by tomaszrykala** [comment]
@devunwired Could you please clarify should `public static Apa102 openLedStrip(Apa102.Direction direction) {... }` be put in place of existing, or added as an overloaded method of existing `public static Apa102 openLedStrip() {... }` ?

**Comment by devunwired** [comment]
My preference would be to keep a single method, and not add any more new methods to the factory.

**Comment by tomaszrykala** [comment]
Implemented.

**Comment by proppy** [comment]
@devunwired maybe we should also modify the `Apa102` driver to be the same?

File: rainbowhat/README.md [hunk 2]
```diff
@@ -92,7 +92,7 @@ segment.close();
 
 ```java
 // Light up the rainbow
-Apa102 ledstrip = RainbowHat.openLedStrip();
+Apa102 ledstrip = RainbowHatopenLedStrip(Apa102.Direction.NORMAL);
```
Comment (devunwired) [inline rainbowhat/README.md]: The dot disappeared.

**Comment by devunwired** [comment]
I don't see that as a necessary update.

File: rainbowhat/README.md (hunk 2 above)
Comment (tomaszrykala) [inline rainbowhat/README.md]: Fixed, thanks

**Comment by tomaszrykala** [comment]
Is there anything else required?

**Comment by mangini** [comment]
It looks weird that the Apa102 driver provides a default direction but the meta rainbowhat driver would enforce specifying one. It seems more sensible if the Apa102 driver exposes a setDirection method, and the user would openLedStrip followed by a call to setDirection if necessary (being NORMAL the default). @proppy @devunwired WDYT?

**Comment by devunwired** [comment]
I could go either way. It doesn't seem a likely parameter to change after init on the Apa102, but I don't see the harm in letting the direction be mutable either. The driver would need to avoid reading the change during the command loop though (right now I think it reads the direction for each led segment).

**Comment by proppy** [comment]
+1 to adding a setDirection call and remove the direction param from the constructors.

**Comment by tomaszrykala** [comment]
@mangini @devunwired @proppy Implemented, please let me know WDYT.

File: apa102/src/test/java/com/google/android/things/contrib/driver/apa102/Apa102Test.java [hunk 3]
```diff
@@ -89,6 +89,45 @@ public void setBrightness_throwsIfTooLarge() throws IOException {
...
+
+        leds.write(colors);
+        Mockito.verify(mSpiDevice).write(Mockito.argThat(BytesMatcher.contains(
+                (byte)(0xE0|brightness),
+                (byte)(colors[2]&0xff), (byte)(colors[2]>>8&0xff), (byte)(colors[2]>>16&0xff),
+                (byte)(0xE0|brightness),
+                (byte)(colors[1]&0xff), (byte)(colors[1]>>8&0xff), (byte)(colors[1]>>16&0xff),
+                (byte)(0xE0|brightness),
+                (byte)(colors[0]&0xff), (byte)(colors[0]>>8&0xff), (byte)(colors[0]>>16&0xff)
+        )), Mockito.eq(headerSize + colors.length*4 + endframeSize));
```
Comment (tomaszrykala) [inline apa102/src/test/java/com/google/android/things/contrib/driver/apa102/Apa102Test.java]: The two blocks of Mockito `verify`'s are reused in two tests each. For the sake of avoiding repetition how do you feel about extracting each block to `assertDirectionNormal()` and `...Reversed()`, or just `assertDirection(Direction direction) { if(direction == ...}` ?

**Pull request merged**